### ISSUE TITLE
PNG with `colors: 8` on image with transparency broken since v0.31.0

### ISSUE THREAD

**Issue by gustavopch** [lovell/sharp#3413]
### Is this a possible bug in a feature of sharp, unrelated to installation?

- [x] Running `npm install sharp` completes without error.
- [x] Running `node -e "require('sharp')"` completes without error.

If you cannot confirm both of these, please open an [installation issue](https://github.com/lovell/sharp/issues/new?labels=installation&template=installation.md) instead.

### Are you using the latest version of sharp?

- [x] I am using the latest version of `sharp` as reported by `npm view sharp dist-tags.latest`.

If you cannot confirm this, please upgrade to the latest version and try again before opening an issue.

If you are using another package which depends on a version of `sharp` that is not the latest, please open an issue against that package instead.

### What is the output of running `npx envinfo --binaries --system --npmPackages=sharp --npmGlobalPackages=sharp`?

```
  System:
    OS: macOS 12.6
    CPU: (8) arm64 Apple M1
    Memory: 161.36 MB / 8.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.13.1 - ~/.asdf/installs/nodejs/16.13.1/bin/node
    Yarn: 1.22.19 - ~/.asdf/installs/nodejs/16.13.1/.npm/bin/yarn
    npm: 8.9.0 - ~/.asdf/installs/nodejs/16.13.1/.npm/bin/npm
    Watchman: 2022.03.14.00 - /opt/homebrew/bin/watchman
  npmPackages:
    sharp: 0.31.1 => 0.31.1
```

### What are the steps to reproduce?

1. Download this PNG file: 
![input](https://user-images.githubusercontent.com/10941958/196579627-f48638f1-8a10-4c9d-8fca-7396c9cadedb.png)
2. Run the code sample with v0.30.7: the generated image will be as expected.
3. Run the code sample with v0.31.0 (or later): the generated image will be blank.

### What is the expected behaviour?

The output image should not be blank. It should be as the one generated with v0.30.7.

### Please provide a minimal, standalone code sample, without other dependencies, that demonstrates this problem

```js
import * as fs from 'node:fs'
import sharp from 'sharp'

const readStream = fs.createReadStream('./input.png')
const pipeline = sharp().png({ colors: 8 })
const writeStream = fs.createWriteStream('./output.png')
readStream.pipe(pipeline).pipe(writeStream)
```

### Please provide sample image(s) that help explain this problem

![image](https://user-images.githubusercontent.com/10941958/196580479-1eb83e85-fe75-4a11-9e50-7df6bf22f6eb.png)

**Comment by lovell** [comment]
Hi, this looks like the same problem as #3395

**Comment by gustavopch** [comment]
> reverting to 0.31.0 works
> – https://github.com/lovell/sharp/issues/3395#issuecomment-1264451788

Last working version for me is v0.30.7.

**Comment by lovell** [comment]
v0.31.0 appears to work for 256 colour "palette" images, but that's due to a side effect of the now-fixed #3357. For 8 colour palettes, you are correct that v0.30.7 is the most recent version that currently works.

(v0.31.0 switched from libpng to libspng for PNG output, hence these issues appearing.)

Let's track this at #3395
